docs/proposals/kubelet-systemd.md

```diff
@@ -0,0 +1,181 @@
+only belong to a single leaf node.
+
+## cgroup hierarchy: orthogonal versus unified hierarchies and cgroupfs single writer
+
+Classical `cgroup` hierarchies were orthogonal per resource group controller, and a process may exist in different parts of the hierarchy.
+
+For example, a process `p1` could exist in each of the following at the same time:
+
+* `/sys/fs/cgroup/cpu/important/`
+* `/sys/fs/cgroup/memory/unimportant/`
+* `/sys/fs/cgroup/cpuacct/unimportant/`
+
+In addition, controllers for one resource group could depend on another in ways that were not always obvious.
+
+For example, the `cpu` controller depends on the `cpuacct` controller yet they were treated separately.
+
+Many found it confusing for a single process to belong to different nodes in the `cgroup` hierarchy across controllers.
+
+The Kernel direction for `cgroup` support is to move toward a unified `cgroup` hierarchy, where the per-controller hierarchies are eliminated in
+favor of hierarchies like the following:
+
+* `/sys/fs/cgroup/important/`
+* `/sys/fs/cgroup/unimportant/`
+
+In a unified hierarchy, a process should only belong to a single node in the `cgroup` tree.
+
+The Kernel direction for `cgroup` management is to promote a single-writer model rather than allowing multiple processes to independently
+write to parts of the file-system.
+
+## cgroup management with systemd and container runtimes
+
+By default, `systemd` defines three slices:
+
+* `system.slice` - contains system and application daemons
+* `user.slice` - user login related processes
+* `machine.slice` - virtual machines or containers that have init systems (not application space containers)
+
+A `slice` corresponds to an inner-node in the `cgroup` hierarchy.
+
+For example, the `system.slice` is represented as `/sys/fs/cgroup/<controller>/system.slice` in the tree.
+
+A `slice` is nested in the hierarchy by it's naming convention.
+
+For example, the `system-foo.slice` is represented as `/sys/fs/cgroup/<controller>/system.slice/system-foo.slice/`.
+
+Finally, to ensure everyone is on the same page, if `docker.service` is associated with `system.slice` and the user launches a container via
+`docker run -m 100M busybox`, a `scope` will be created because the process was not launched by `systemd` itself in the `slice` associated with the launching daemon.
+
+For example:
+
+`/sys/fs/cgroup/system.slice/docker-40c5264f71043fe15193603a05c7bf2a6b7f42b06951a7b59357cc7c2a380f04.scope`
+
+## kubelet cgroup driver
+
+The `kubelet` manipulates the `cgroup` file-system directly during initial bootstrapping.
+
+This proposal proposes that the `kubelet` accept a new flag:
+
+* `--cgroup-driver="": driver to use for cgroup manipulation by the kubelet.  Possible values: cgroupfs, systemd`
+
+The `ContainerManager` interface in the `kubelet` should support an implementation for `cgroupfs` or `systemd`.
+
+Effort to share code across implementations should be made or separated out.
```

Comment:
not sure what "should be made or separated out" means - perhaps just "Shared code across implementations should be split into a common location"